internal/ctrlflow/ctrlflow.go

```diff
@@ -0,0 +1,189 @@
+package ctrlflow
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"log"
+	mathrand "math/rand"
+	"strconv"
+	"strings"
+
+	"golang.org/x/tools/go/ast/astutil"
+	"golang.org/x/tools/go/ssa"
+	ah "mvdan.cc/garble/internal/asthelper"
+	"mvdan.cc/garble/internal/ssa2ast"
+)
+
+const (
+	FileName = "GARBLE_controlflow.go"
+
+	directiveName = "//garble:controlflow"
+	importPrefix  = "___garble_import"
+
+	defaultSplit  = 0
+	defaultJunk   = 0
+	defaultPasses = 1
+)
+
+type directiveParamMap map[string]string
+
+func (m directiveParamMap) GetInt(name string, def int) int {
+	rawVal, ok := m[name]
+	if !ok {
+		return def
+	}
+
+	val, err := strconv.Atoi(rawVal)
+	if err != nil {
+		panic(fmt.Errorf("invalid flag %s format: %v", name, err))
+	}
+	return val
+}
+
+// parseDirective parses a directive string and returns a map of directive parameters.
+// Each parameter should be in the form "key=value" or "key"
+func parseDirective(directive string) (directiveParamMap, bool) {
+	if !strings.HasPrefix(directive, directiveName) {
+		return nil, false
+	}
+
+	fields := strings.Fields(directive)
+	if len(fields) <= 1 {
+		return nil, true
+	}
+	m := make(map[string]string)
+	for _, v := range fields[1:] {
+		params := strings.SplitN(v, "=", 2)
+		if len(params) == 2 {
+			m[params[0]] = params[1]
+		} else {
+			m[params[0]] = ""
+		}
+	}
+	return m, true
+}
+
+// Obfuscate obfuscates control flow of all methods with directive using control flattening.
+// All obfuscated methods are removed from the original file and moved to the new one.
+// Obfuscation can be customized by passing parameters from the directive, example:
+//
+// //garble:controlflow passes=1 junk=0 split=0
+// func someMethod() {}
+//
+// passes - controls number of passes of control flow flattening. Have exponential complexity and more than 3 passes are not recommended in most cases.
+// junk - controls how many junk jumps are added. It does not affect final binary by itself, but together with flattening linearly increases complexity.
+// split - controls number of times largest block must be splitted. Together with flattening improves obfuscation of long blocks without branches.
+func Obfuscate(fset *token.FileSet, ssaPkg *ssa.Package, files []*ast.File, obfRand *mathrand.Rand) (newFile *ast.File, affectedFiles []*ast.File, err error) {
+	var ssaFuncs []*ssa.Function
+	var ssaParams []directiveParamMap
+
+	for _, file := range files {
+		affected := false
+		ast.Inspect(file, func(node ast.Node) bool {
```

Comment:
don't use AST walking to iterate over func decls; it's easier and cheaper to simply loop over `file.Decls`.